### DISCUSSION TITLE
Deployment of Next.js Prisma to Vercel - 500

### DISCUSSION THREAD

**Discussion by GoodFaithParadigm8** [prisma/prisma#3002] [Q&A] [answered]
Getting a 500 when deploying to Vercel. Doesn't seem like the database env is being read from variable stored on vercel. Tried just an env and a global secret. In the vercel config file tried root and nested configuration to (possibly) scope to the prisma file. What is needed to push this out to production?

**Comment by GoodFaithParadigm8** [comment]
I tried pushing my .env file in the prisma folder with a commit also with no avail.

**Comment by 2color** [comment, marked as answer]
Did you follow the details of step 6 in the [Verel deployment guide](https://www.prisma.io/docs/guides/deployment/deploying-to-vercel#6-define-vercel-secret-and-expose-to-functions)?
Can you share your `vercel.json` file (and verify if the secret exists in vercel with `vercel secrets ls`)
Which version of Prisma are you using?

Generally, it's recommended to keep secrets out of the repository and load them at runtime. [Vercel has its own mechanism](https://vercel.com/docs/v2/build-step#environment-variables) for setting secrets and passing them as environment variables. So there's no need for an `.env` file (except for [local development](https://vercel.com/docs/v2/build-step#development-environment-variables))

**Reply by GoodFaithParadigm8** [reply]
Thanks @2color. I did follow step 6 in the deployment guide. I deleted the vercel.json file when I push the .env file into the repository. I am using prisma 2.13. I understand the security purposes of keeping a .env file out of the repository. There is nothing in the database that is sensitive and it is spun up purely for experimental development purposes. It will get deleted once I settle on a development platform.

This was the vercel.json file.

```
{
    "version": 2,
    "env": {
        "DATABASE_URL": "@database_url"
    },
    "build": { 
        "env": {
            "DATABASE_URL": "@database_url"
        }
    }
}   
```

**Reply by 2color** [reply]
Can you share the logs of the function? That should provide more insight into what's happening when the functions get called.

**Reply by GoodFaithParadigm8** [reply]
I tried clearing all the secrets that I had stored in Vercel. I've previously had issues (presumably) with the provider side having some cache mechanism (not just now or vercel) that makes clearing out any data store troublesome. I gave it some time before saving another secret. Instead of pushing to git I tried to deloy with `vercel -e NODE_ENV=production -e DATABASE_URL=@database_url` and got an error.

`Error! Environment Variable "DATABASE_URL" references Secret "sec_4x4Fn0qWSkYK2uEBxT76sGdR", which was deleted.
`

Next I tried pushing to the git repository with the same vercel.json file. Still a 500. There were no logs, but it did parse all the functions so I curl'd one of them.

Here is the log.
```
[GET] /api/drafts
21:34:13:12
2020-07-14T02:34:15.060Z	91bea408-ab7c-4112-8608-28f35a10e05a	ERROR	PrismaClientUnknownRequestError: Error in connector: Error querying the database: db error: FATAL: no pg_hba.conf entry for host "x.x.x.x", user "username", database "prisma1", SSL on
    at PrismaClientFetcher.request (/var/task/node_modules/@prisma/client/runtime/index.js:1:185866)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Reply by GoodFaithParadigm8** [reply]
I also just tried this `vercel -e NODE_ENV=production -e DATABASE_URL="MyDatabaseConnectionString"` and still got a 500.

**Reply by 2color** [reply]
When using the command `vercel -e NODE_ENV=production -e DATABASE_URL=@database_url` I believe that you have to pass the value of  `DATABASE_URL` rather than a reference to a secret.

Can you give it a go with the actual DATABASE_URL, e.g. `vercel -e NODE_ENV=production -e DATABASE_URL="postgresql://__USER__:__PASSWORD__@__HOST__/__DATABASE__"`

When running the application locally with `vercel dev` are you able to call the functions?

You could debug this by logging `process.env.DATABASE_URL` in your functions before instantiating the Prisma Client to see what the application gets at runtime.

**Reply by GoodFaithParadigm8** [reply]
I did put my actual database string when I tried `vercel -e NODE_ENV=production -e DATABASE_URL="MyDatabaseConnectionString"`

I am also not able to run it locally. I get the error - `Error! Env var "DATABASE_URL" is not defined in `.env.build` file`. 

I've tried saving the env variable several different ways: storing it manually in the vercel console, storing it as a variable in the console and then referencing it when saving secrets with the vercel cli, and initiating the variable running the build with the cli.

When I run `vercel env pull` I get a blank env variable, but when I call one of the functions that are parsed the error makes reference to my username and the fact that it is run over ssl as you can see above. 🤷

**Reply by 2color** [reply]
I'd first try upgrading Prisma to `2.2.2`. We recently [merged a bug fix](https://github.com/prisma/prisma/issues/2971) related to loading environment variables from `.env` files.

It sounds like the problem is with `process.env.DATABASE_URL` being available at application runtime. I'd first suggest figuring out whether this is related to Prisma or whether it's simply because `process.env.DATABASE_URL` is not correctly set. 

My hunch is that it's probably defined in more than one place (as an env var or by reading `.env`) and it gets overwritten by an empty value.

Prisma provides native support for .env files if the .env file is located in the same directory as your Prisma schema file. 

I'd test running the application locally first with node directly (not with `vercel dev`) and then seeing if it pops up when deployed.

Besides that, I'd suggest maybe deleting the Vercel project and re-creating it.

**Reply by GoodFaithParadigm8** [reply]
When storing the secret with or without ~~string~~ quotations, I get a bash error for the `!` character. I resolve this error by escaping the `!`. It seems though, after console logging one of the api routes as you suggested, that the connection string is saved with the escape character, thus not connecting to the database.

**Reply by GoodFaithParadigm8** [reply]
Using the HTML ASCII characters seems to fix the issue with `!`. However, now calling the function that has the console log I get an invalid port error. 

```
[GET] /api/feed
13:22:41:27
postgresql://xxxx@xxxx**&#33**@gxxx:5432/prisma1?schema=public&sslmode=require
2020-07-14T18:22:41.356Z	25cb7d4e-8e40-47fe-83ab-ca20af7d15bb	ERROR	PrismaClientInitializationError: Error parsing connection string: invalid port number
    at PrismaClientFetcher.request (/var/task/node_modules/@prisma/client/runtime/index.js:1:185969)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Reply by GoodFaithParadigm8** [reply]
>I'd test running the application locally first with node directly (not with vercel dev) and then seeing if it pops up when deployed. Besides that, I'd suggest maybe deleting the Vercel project and re-creating it.

The project runs fine locally. I just wanted to see what behavior was different pushing it out to production. Appreciate your help. A couple of suggestions you gave helped troubleshoot this issue better. I see using the `!` character with vercel and next may not be too wise. It doesn't appear to be a Prisma issue.

**Reply by 2color** [reply]
Glad it worked out. 

For future reference, can you verify that the problem was a `!` character in your PostgreSQL URI that caused? Was it part of the username or password?

**Reply by GoodFaithParadigm8** [reply]
@2color Yes it was the username password and storing that secret in Vercel.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
